# Define UI for data upload app ----
library("shiny", quietly = TRUE)



quick_start_page <- function() {
  bslib::card(
    class = "border-danger",
    bslib::card_header(class = "bg-danger text-light", "Getting Started"),
    h3(tags$b(span("Quick Start", style = "text-decoration:underline"))),
    tags$ul(
      tags$li("Upload your ", tags$b("proteinGroups.txt "), "generated by MaxQuant."),
      tags$li("Upload your ", tags$b(" experimental design "), "table. "),
      tags$li(tags$b("Optional: "), "Adjust the p-value cut-off, the log2 fold change cut-off,
                                the imputation type, FDR correction method and/or number of clusters in heatmap
                                in the", tags$b("Advanced Options")),
      tags$li("Press ", tags$b("'Start Analysis' ")),
      tags$li(tags$b("Note: "), " The experimental design file is not the", tags$b("'mqpar.xml' "), "file
                        from MaxQuant.")
    ),
    br(),
    HTML('<center><img src="/static/img/LFQ_analyst.svg" width="500px"></center>')
  )
}

# downloadTable
# downloadButton
# significantBox
# downloadreport
top_row <- function() {
  bslib::layout_columns(
    col_widths = c(4, 5, 3),
    bslib::card(
      bslib::layout_columns(
        selectizeInput(
          "dataset",
          "Choose a dataset to save",
          choice = c(
            "Results", "Original_matrix",
            "Imputed_matrix",
            "Full_dataset"
          ),
          options = list(dropdownParent = "body")
        ),
        downloadButton("downloadData", "Save", class = "mt-4")
      )
    ),
    uiOutput("significantBox"),
    bslib::card(
      downloadButton("downloadReport", "Download Report", class = "mt-4")
    )
  )
}

# contents
# original
results_table <- function() {
  bslib::card(
    class = "border-success",
    bslib::card_header(class = "bg-success", "LFQ Results Table"),
    DT::dataTableOutput("contents"),
    actionButton("original", "Refresh Table")
  )
}
# volcano_cntrst
# fontsize
# check_names
# p_adj
# volcano protein_brush protein_click
# downloadVolcano
# resetPlot
# heatmap
# cluster_number
# downloadCluster
# download_hm_svg
# protein_plot_type
# protein_plot
# downloadProtein
results_plots <- function() {
  bslib::navset_card_underline(
    title = "Result Plots",
    bslib::nav_panel(
      title = "Volcano plot",
      bslib::layout_columns(
        # col_widths = c(4, 4, 4),
        style = htmltools::css(grid_template_columns = "1fr 1fr 1fr"),
        uiOutput("volcano_cntrst"),
        numericInput("fontsize",
          "Font size",
          min = 0, max = 8, value = 4
        ),
        bslib::card(
          checkboxInput("check_names",
            "Display names",
            value = FALSE
          ),
          checkboxInput("p_adj",
            "Adjusted p values",
            value = FALSE
          )
        )
      ),
      tags$p("Select protein from LFQ Results Table to highlight on the plot OR
                                                  drag the mouse on plot to show expression of proteins in Table"),
      bslib::card(
        plotOutput("volcano",
          height = 600,
          # hover = "protein_hover"),
          # ),
          # click = "protein_click"),
          brush = "protein_brush",
          click = "protein_click"
        ),
        downloadButton("downloadVolcano", "Save Highlighted Plot"),
        actionButton("resetPlot", "Clear Selection")
      )
    ),
    bslib::nav_panel(
      title = "Heatmap",
      plotOutput("heatmap", height = 600),
      bslib::card(
        numericInput("cluster_number",
          "Cluster to download",
          min = 1, max = 6, value = 1
        ),
        width = 6
      ),
      bslib::card(downloadButton("downloadCluster", "Save Cluster"),
        downloadButton("download_hm_svg", "Save svg"),
        width = 5
      )
      # # align save button
      # tags$style(type = "text/css", "#downloadCluster {margin-top: 25px;}"),
      # tags$style(type = "text/css", "#download_hm_svg {margin-top: 25px;}")
    ),
    bslib::nav_panel(
      title = "Protein Plot",
      bslib::card(
        radioButtons("protein_plot_type",
          "Plot type",
          choices = c(
            "Box Plot" = "boxplot",
            "Violin Plot" = "violin",
            "Interaction Plot" = "interaction",
            "Intensity Plot" = "dot"
          ),
          selected = "boxplot",
          inline = TRUE
        )
      ),
      tags$p("Select one or more rows from LFQ Results Table to plot individual
                                                  protein intesities across conditions and replicates"),
      bslib::card(
        plotOutput("protein_plot"),
        downloadButton("downloadProtein", "Download Plot")
      )
    )
  )
}

# pca_plot
# download_pca_svg
# sample_corr_plot
# download_corr_svg
# sample_cvs_plot
# download_cvs_svg
# numbers_plot
# download_num_svg
# coverage_plot
# download_cov_svg
# normalization_plot
# download_norm_svg
# missval_plot
# download_missval_svg
# imputation_plot
# download_imp_svg
qc_plots <- function() {
  bslib::navset_card_underline(
    # title = "QC Plots",
    bslib::nav_panel(
      title = "PCA Plot",
      plotOutput("pca_plot", height = 600),
      downloadButton("download_pca_svg", "Save svg")
    ),
    bslib::nav_panel(
      title = "Sample Correlation",
      plotOutput("sample_corr_plot", height = 600),
      downloadButton("download_corr_svg", "Save svg")
    ),
    bslib::nav_panel(
      title = "Sample CVs",
      plotOutput("sample_cvs_plot", height = 600),
      downloadButton("download_cvs_svg", "Save svg")
    ),
    bslib::nav_panel(
      title = "Protein Numbers",
      plotOutput("numbers_plot", height = 600),
      downloadButton("download_num_svg", "Save svg")
    ),
    bslib::nav_panel(
      title = "Sample coverage",
      plotOutput("coverage_plot", height = 600),
      downloadButton("download_cov_svg", "Save svg")
    ),
    bslib::nav_panel(
      title = "Normalization",
      plotOutput("normalization_plot", height = 600),
      downloadButton("download_norm_svg", "Save svg")
    ),
    bslib::nav_panel(
      title = "Missing values - Heatmap",
      plotOutput("missval_plot", height = 600),
      downloadButton("download_missval_svg", "Save svg")
    ),
    bslib::nav_panel(
      title = "Imputation",
      plotOutput("imputation_plot", height = 600),
      downloadButton("download_imp_svg", "Save svg")
    ),
    bslib::nav_spacer(),
    bslib::nav_item("QC Plots")
  )
}

# contrast_placeholder
# go_database
# go_analysis
# spinner_go
# downloadGO
# contrast_1_placeholder
# pathway_database
# pathway_analysis
# spinner_pa
# downloadPA
enrichment_box <- function() {
  bslib::navset_card_underline(
    title = "Enrichment",
    bslib::nav_panel(
      title = "Gene Ontology",
      bslib::layout_columns(
        uiOutput("contrast_placeholder"),
        selectizeInput(
          "go_database", "GO database:",
          choices = c(
            "Molecular Function" = "GO_Molecular_Function_2021",
            "Cellular Component" = "GO_Cellular_Component_2021",
            "Biological Process" = "GO_Biological_Process_2021"
          ),
          options = list(dropdownParent = "body")
        ),
        actionButton("go_analysis", "Run Enrichment", class = "mt-4")
      ),
      bslib::card(uiOutput("spinner_go")),
      downloadButton("downloadGO", "Download Table")
    ),
    bslib::nav_panel(
      title = "Pathway enrichment",
      bslib::layout_columns(
        uiOutput("contrast_1_placeholder"),
        selectizeInput(
          "pathway_database", "Pathway database:",
          choices = c(
            "KEGG" = "KEGG_2021_Human",
            "Reactome" = "Reactome_2022"
          ),
          options = list(dropdownParent = "body")
        ),
        actionButton("pathway_analysis", "Run Enrichment"),
      ),
      bslib::card(uiOutput("spinner_pa"), height = 400),
      downloadButton("downloadPA", "Download Table", class = "mt-4")
    )
  )
}



analysis_tab <- function() {
  tagList(
    div(
      id = "quickstart_info", quick_start_page()
    ),
    shinyjs::hidden(
      div(
        id = "analysis_id",
        top_row(),
        bslib::layout_columns(
          results_table(),
          results_plots()
        ),
        bslib::layout_columns(
          qc_plots(),
          enrichment_box()
        )
      )
    )
  )
}


# maxquant_file
# exp_design_file
# analyze
# p_value
# log_fold_change
# paired
# imputation
# fdr_correction
# single_peptide
# k_number
sidebar_panel <- function() {
  bslib::layout_sidebar(
    title = "Analysis",
    sidebar = bslib::sidebar(
      width = "20rem",
      bslib::accordion(
        bslib::accordion_panel(
          title = "Analysis", icon = icon("flask"),
          open = TRUE,
          fileInput("maxquant_file",
            "Upload MaxQuant ProteinGroups.txt",
            accept = c(
              "text/csv",
              "text/comma-separated-values,text/plain",
              ".csv"
            )
          ),
          fileInput("exp_design_file",
            "Upload Experimental Design Matrix",
            accept = c(
              "text/csv",
              "text/comma-separated-values,text/plain",
              ".csv"
            )
          ),
          tags$hr(),
          actionButton("analyze", "Start Analysis")
        ),
        bslib::accordion_panel(
          title = "Advanced Options",
          icon = icon("cogs"),
          numericInput("p_value",
            "Adjusted p-value cutoff",
            min = 0.0001, max = 0.1, value = 0.05
          ),
          numericInput("log_fold_change",
            "Log2 fold change cutoff",
            min = 0, max = 10, value = 1
          ),
          checkboxInput(
            "paired",
            "Paired test", FALSE
          ),
          radioButtons("imputation",
            "Imputation type",
            choices = c("Perseus-type" = "man", MsCoreUtils::imputeMethods())[1:9],
            selected = "man"
          ),
          radioButtons("fdr_correction",
            "Type of FDR correction",
            choices = c(
              "Benjamini Hochberg" = "BH",
              "t-statistics-based" = "fdrtool"
            ), selected = "BH"
          ),
          checkboxInput(
            "single_peptide",
            "Include single peptide identifications", FALSE
          ),
          numericInput("k_number",
            "Number of clusters in heatmap",
            min = 1, max = 20, value = 6
          )
        )
      )
    ),
    analysis_tab()
  )
}


link_LFQ <- tags$a(icon("chart-line"), "LFQ-Analyst",
  href = "https://bioinformatics.erc.monash.edu/apps/LFQ-Analyst/",
  target = "_blank"
)

ui <- shinyUI({
  bslib::page_navbar(
    title = tags$span("LFQ"),
    theme = bslib::bs_theme(version = 5, font_scale = .8),
    navbar_options = bslib::navbar_options(
      bg = "#0062cc"
    ),
    id = "page",
    window_title = "LFQ",
    fillable = FALSE,
    header = tagList(
      shinydisconnect::disconnectMessage(),
      tags$head(
        # tags$link(rel = "icon", href = ICON),
        # tags$link(rel = "stylesheet", type = "text/css", href = "/static/css/custom.css?v=1.5"),
        tags$style(HTML("
          root: {
            --bslib-spacer: 0px;
          }
          .card,.well {
            --bs-card-border-radius: 0px;
            --bs-card-inner-border-radius: 0px;
          }

        ")),
        shinyjs::useShinyjs() # imp to use shinyjs functions
      )
    ),
    bslib::nav_panel(
      title = "LFQ", value = "lfq", icon = icon("list"),
      sidebar_panel()
    ),
    bslib::nav_spacer(),
    bslib::nav_item(link_LFQ)
    # footer = tagList(
    #   tags$p("Supported by: Monash Proteomics and Metabolomics Platform & Monash Bioinformatics Platform,
    #           Monash University"),
    #   #shiny.info::version(position = "bottom right")
    # )
  )
})
