# Define UI for data upload app ----
library("shiny", quietly = TRUE)
# https://stackoverflow.com/questions/31794702/r-shiny-dashboard-tabitems-not-apparing
convertMenuItem <- function(mi, tabName) {
  mi$children[[1]]$attribs["data-toggle"] <- "tab"
  mi$children[[1]]$attribs["data-value"] <- tabName
  mi
}


quick_start_page <- function() {
  fluidPage(
    shinydashboard::box(
      title = "Getting Started",
      h3(tags$b(span("Quick Start", style = "text-decoration:underline"))),
      tags$ul(
        tags$li("Upload your ", tags$b("proteinGroups.txt "), "generated by MaxQuant."),
        tags$li("Upload your ", tags$b(" experimental design "), "table. "),
        tags$li(tags$b("Optional: "), "Adjust the p-value cut-off, the log2 fold change cut-off,
                                the imputation type, FDR correction method and/or number of clusters in heatmap
                                in the", tags$b("Advanced Options")),
        tags$li("Press ", tags$b("'Start Analysis' ")),
        tags$li(tags$b("Note: "), " The experimental design file is not the", tags$b("'mqpar.xml' "), "file
                        from MaxQuant.")
      ),
      br(),
      HTML('<center><img src="/static/img/LFQ_analyst.svg" width="500px"></center>'),
      width = 12,
      solidHeader = TRUE,
      status = "danger"
    )
  )
}

# downloadTable
# downloadButton
# significantBox
# downloadreport
top_row <- function() {
  fluidRow(
    shinydashboard::box(
      column(6, uiOutput("downloadTable"), offset = 1),
      column(4, uiOutput("downloadButton")), # make the button on same line
      width = 4
    ),
    shinydashboard::infoBoxOutput("significantBox", width = 4),
    shinydashboard::box(
      column(5, uiOutput("downloadreport")), # offset for dist between buttons
      # tags$br(),
      # column(5,uiOutput('downloadPlots')),
      width = 4
    )
  )
}

# contents
# original
results_table <- function() {
  shinydashboard::box(
    title = "LFQ Results Table",
    DT::dataTableOutput("contents"),
    #  actionButton("clear", "Deselect Rows"),
    actionButton("original", "Refresh Table"),
    width = 6,
    status = "success",
    # color=""
    solidHeader = TRUE
  )
}
# volcano_cntrst
# fontsize
# check_names
# p_adj
# volcano protein_brush protein_click
# downloadVolcano
# resetPlot
# heatmap
# cluster_number
# downloadCluster
# download_hm_svg
# protein_plot_type
# protein_plot
# downloadProtein
results_plots <- function() {
  shinydashboard::box(
    width = 6,
    collapsible = TRUE,
    # status="primary",
    # solidHeader=TRUE,
    shinydashboard::tabBox(
      title = "Result Plots",
      width = 12,
      tabPanel(
        title = "Volcano plot",
        fluidRow(
          shinydashboard::box(uiOutput("volcano_cntrst"), width = 5),
          shinydashboard::box(
            numericInput("fontsize",
              "Font size",
              min = 0, max = 8, value = 4
            ),
            width = 3
          ),
          shinydashboard::box(
            checkboxInput("check_names",
              "Display names",
              value = FALSE
            ),
            checkboxInput("p_adj",
              "Adjusted p values",
              value = FALSE
            ),
            width = 4
          ),
          tags$p("Select protein from LFQ Results Table to highlight on the plot OR
                                                  drag the mouse on plot to show expression of proteins in Table")
          # Add text line
          # tags$p("OR"),
          #  tags$p("Drag the mouse on plot to show expression of proteins in Table")
        ),
        fluidRow(
          plotOutput("volcano",
            height = 600,
            # hover = "protein_hover"),
            # ),
            # click = "protein_click"),
            brush = "protein_brush",
            click = "protein_click"
          ),
          downloadButton("downloadVolcano", "Save Highlighted Plot"),
          actionButton("resetPlot", "Clear Selection")
          # )),
        )
      ),
      tabPanel(
        title = "Heatmap",
        fluidRow(
          plotOutput("heatmap", height = 600)
        ),
        fluidRow(
          shinydashboard::box(
            numericInput("cluster_number",
              "Cluster to download",
              min = 1, max = 6, value = 1
            ),
            width = 6
          ),
          shinydashboard::box(downloadButton("downloadCluster", "Save Cluster"),
            downloadButton("download_hm_svg", "Save svg"),
            width = 5
          ),
          # align save button
          tags$style(type = "text/css", "#downloadCluster {margin-top: 25px;}"),
          tags$style(type = "text/css", "#download_hm_svg {margin-top: 25px;}")
        )
      ),
      tabPanel(
        title = "Protein Plot",
        fluidRow(
          shinydashboard::box(
            radioButtons("protein_plot_type",
              "Plot type",
              choices = c(
                "Box Plot" = "boxplot",
                "Violin Plot" = "violin",
                "Interaction Plot" = "interaction",
                "Intensity Plot" = "dot"
              ),
              selected = "boxplot",
              inline = TRUE
            ),
            width = 12
          ),
          tags$p("Select one or more rows from LFQ Results Table to plot individual
                                                  protein intesities across conditions and replicates")
        ),
        fluidRow(
          plotOutput("protein_plot"),
          downloadButton("downloadProtein", "Download Plot")
        )
      )
    )
  )
}

# pca_plot
# download_pca_svg
# sample_corr_plot
# download_corr_svg
# sample_cvs_plot
# download_cvs_svg
# numbers_plot
# download_num_svg
# coverage_plot
# download_cov_svg
# normalization_plot
# download_norm_svg
# missval_plot
# download_missval_svg
# imputation_plot
# download_imp_svg
qc_plots <- function() {
  shinydashboard::tabBox(
    title = "QC Plots", width = 12,
    tabPanel(
      title = "PCA Plot",
      plotOutput("pca_plot", height = 600),
      downloadButton("download_pca_svg", "Save svg")
    ),
    tabPanel(
      title = "Sample Correlation",
      plotOutput("sample_corr_plot", height = 600),
      downloadButton("download_corr_svg", "Save svg")
    ),
    tabPanel(
      title = "Sample CVs",
      plotOutput("sample_cvs_plot", height = 600),
      downloadButton("download_cvs_svg", "Save svg")
    ),
    tabPanel(
      title = "Protein Numbers",
      plotOutput("numbers_plot", height = 600),
      downloadButton("download_num_svg", "Save svg")
    ),
    tabPanel(
      title = "Sample coverage",
      plotOutput("coverage_plot", height = 600),
      downloadButton("download_cov_svg", "Save svg")
    ),
    tabPanel(
      title = "Normalization",
      plotOutput("normalization_plot", height = 600),
      downloadButton("download_norm_svg", "Save svg")
    ),
    tabPanel(
      title = "Missing values - Heatmap",
      plotOutput("missval_plot", height = 600),
      downloadButton("download_missval_svg", "Save svg")
    ),
    tabPanel(
      title = "Imputation",
      plotOutput("imputation_plot", height = 600),
      downloadButton("download_imp_svg", "Save svg")
    )
  )
}

# contrast
# go_database
# go_analysis
# spinner_go
# downloadGO
# contrast_1
# pathway_database
# pathway_analysis
# spinner_pa
# downloadPA
enrichment_box <- function() {
  shinydashboard::tabBox(
    title = "Enrichment", width = 12,
    tabPanel(
      title = "Gene Ontology",
      fluidRow(
        column(
          6,
          uiOutput("contrast")
        ),
        column(
          6,
          selectInput(
            "go_database", "GO database:",
            c(
              "Molecular Function" = "GO_Molecular_Function_2021",
              "Cellular Component" = "GO_Cellular_Component_2021",
              "Biological Process" = "GO_Biological_Process_2021"
            )
          )
        ),
        column(12, actionButton("go_analysis", "Run Enrichment")),
        column(
          12,
          shinydashboard::box(width = 12, uiOutput("spinner_go"), height = 400)
        ),
        column(12, downloadButton("downloadGO", "Download Table"))
      )
    ),
    tabPanel(
      title = "Pathway enrichment",
      fluidRow(
        column(
          6,
          uiOutput("contrast_1")
        ),
        column(
          6,
          selectInput(
            "pathway_database", "Pathway database:",
            c(
              "KEGG" = "KEGG_2021_Human",
              "Reactome" = "Reactome_2022"
            )
          )
        ),
        column(12, actionButton("pathway_analysis", "Run Enrichment")),
        column(
          12,
          shinydashboard::box(width = 12, uiOutput("spinner_pa"), height = 400)
        ),
        column(12, downloadButton("downloadPA", "Download Table"))
      )
    )
  ) # Tab box close
}


# maxquant_file
# exp_design_file
# analyze
# p_value
# log_fold_change
# paired
# imputation
# fdr_correction
# single_peptide
# k_number
sidebar <- function() {
  shinydashboard::sidebarMenu(
    id = "tabs_selected",
    convertMenuItem(shinydashboard::menuItem("Analysis",
      tabName = "analysis", icon = icon("flask"),
      startExpanded = TRUE,
      # menuItem("Input Files", tabName="file", icon=icon("file"), #selected = TRUE,
      fileInput("maxquant_file",
        "Upload MaxQuant ProteinGroups.txt",
        accept = c(
          "text/csv",
          "text/comma-separated-values,text/plain",
          ".csv"
        )
      ),
      fileInput("exp_design_file",
        "Upload Experimental Design Matrix",
        accept = c(
          "text/csv",
          "text/comma-separated-values,text/plain",
          ".csv"
        )
      ),
      tags$hr(),
      actionButton("analyze", "Start Analysis"),
      tags$hr(),
      shinydashboard::menuItem("Advanced Options",
        tabName = "advanced", icon = icon("cogs"),
        numericInput("p_value",
          "Adjusted p-value cutoff",
          min = 0.0001, max = 0.1, value = 0.05
        ),
        numericInput("log_fold_change",
          "Log2 fold change cutoff",
          min = 0, max = 10, value = 1
        ),
        checkboxInput(
          "paired",
          "Paired test", FALSE
        ),
        radioButtons("imputation",
          "Imputation type",
          choices = c("Perseus-type" = "man", MsCoreUtils::imputeMethods())[1:9],
          selected = "man"
        ),
        radioButtons("fdr_correction",
          "Type of FDR correction",
          choices = c(
            "Benjamini Hochberg" = "BH",
            "t-statistics-based" = "fdrtool"
          ), selected = "BH"
        ),
        checkboxInput(
          "single_peptide",
          "Include single peptide identifications", FALSE
        ),
        numericInput("k_number",
          "Number of clusters in heatmap",
          min = 1, max = 20, value = 6
        )
      ),
      tags$hr()
    ), tabName = "analysis")
  )
}


analysis_tab <- function() {
  shinydashboard::tabItem(
    tabName = "analysis",
    div(
      id = "quickstart_info",
      quick_start_page()
    ),
    shinyjs::hidden(div(
      id = "downloadbox",
      top_row()
    )), # close div and first row

    # align save button
    tags$style(type = "text/css", "#downloadButton { width:100%; margin-top: 25px;}"),
    tags$style(type = "text/css", "#downloadreport { width:100%; vertical-align- middle; margin-top: 25px;
                 margin-bottom: 25px;}"),
    # tags$style(type='text/css', "#downloadPlots { width:100%; margin-top: 25px;}"),

    tags$br(), # Blank lines
    tags$br(),

    ## Data table and result plots box
    fluidRow(
      shinyjs::hidden(div(
        id = "results_tab",
        results_table(),
        results_plots()
      ))
    ),

    ## QC Box
    fluidRow(
      shinyjs::hidden(div(
        id = "qc_tab",
        column(
          width = 6,
          qc_plots()
        ),
        column(
          width = 6,
          enrichment_box()
        )
      ))
    )
  )
}

ui <- shinyUI(
  shinydashboard::dashboardPage(
    skin = "blue",
    shinydashboard::dashboardHeader(title = "LFQ-Analyst"),
    # disable = TRUE),# Disable title bar
    shinydashboard::dashboardSidebar(
      sidebar()
    ),

    shinydashboard::dashboardBody(
      shinyjs::useShinyjs(), # imp to use shinyjs functions
      tags$head(
        tags$link(rel = "stylesheet", type = "text/css", href = "/static/css/custom.css")
      ),
      tags$style(
        ".box {
        border-top: none;
        box-shadow: 0 0px 0px rgb(0 0 0 / 10%);
        }"
      ),


      shinydashboard::tabItems(
        analysis_tab()
      ),
      tags$footer(
        tags$p("Supported by: Monash Proteomics and Metabolomics Platform & Monash Bioinformatics Platform,
               Monash University"),
        align = "right"
      ),
      shiny.info::version(position = "bottom right")
    )
  )
)
